storage/zynq7000-flash/flashdrv.c

```diff
@@ -398,126 +357,424 @@ ssize_t flashdrv_read(addr_t offs, void *buff, size_t len, time_t timeout)
 }
 
 
-ssize_t flashdrv_write(addr_t offs, const void *buff, size_t len)
+/* Block device interface */
+
+static int _flashdrv_sync(unsigned int id, size_t regStart)
 {
+	uint8_t *src;
+	addr_t dst;
 	ssize_t res;
-	time_t timeout;
-	size_t chunkSz = 0, freeSz = 0, saveSz = 0;
-	static const uint32_t timeoutFactor = 0x100;
-
-	uint32_t regID = 0, sectID = 0;
-	size_t sectSz, regStart;
+	unsigned int i, pgNb;
 
+	size_t pageSz, sectSz;
 	const flash_cfi_t *cfi = &fdrv_common.info.cfi;
 
-	if (len == 0)
-		return 0;
+	if (fdrv_common.regs[id].buffPos == 0) {
+		return EOK;
+	}
+
+	sectSz = CFI_SIZE_SECTION(cfi->regs[id].size);
+	pageSz = CFI_SIZE_PAGE(cfi->pageSize);
+	pgNb = sectSz / pageSz;
+
+	for (i = 0; i < pgNb; ++i) {
+		dst = regStart + fdrv_common.regs[id].sectID * sectSz + i * pageSz;
+		src = fdrv_common.regs[id].buff + i * pageSz;
+
+		res = _flashdrv_pageProgram(id, dst, src, pageSz);
+		if (res < 0) {
+			return res;
+		}
+	}
+
+	fdrv_common.regs[id].sectID = (uint32_t)-1;
+	fdrv_common.regs[id].buffPos = 0;
+
+	return EOK;
+}
 
-	if (buff == NULL || (offs + len) > CFI_SIZE_FLASH(cfi->chipSize))
+static ssize_t flashdrv_blkRead(struct _storage_t *strg, off_t start, void *data, size_t size)
+{
+	ssize_t res;
+
+	if ((start + size) > strg->size) {
 		return -EINVAL;
+	}
 
-	while (saveSz < len) {
-		offs += chunkSz;
+	mutexLock(fdrv_common.regs[strg->dev->id].lock);
+	res = _flashdrv_read(strg->dev->id, start + strg->start, data, size, TIMEOUT_CMD_MS * size);
```

Comment:
IMHO it would be better to pass timeout per cmd (or even don't pass timeout at all), _flashdrv_read knows about TIMEOUT_CMD_MS and size and can do this operation by itself

Reply:
Thanks for suggestion. I have removed the timeout argument from `_flashdrv_read`.